lib/ss-search.c

```diff
@@ -36,121 +36,123 @@
  *
-*/
+ */
 
+// define this internally, it's a "private" function...
 void ss_search_reset(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, int64_t now);
 
 void ss_search_reset(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, int64_t now)
 {
-	// Handy pointers to the cc struct
-	uint64_t* delv = cc->ss_state.search.delv_bins;
-	int64_t* bin_end = &cc->ss_state.search.bin_end;
-	uint32_t* bin_time = &cc->ss_state.search.bin_time;
-	uint32_t* bin_rounds = &cc->ss_state.search.bin_rounds;
+    // Handy pointers to the cc struct
+    QUICLY_SEARCH_BIN_TYPE* delv = cc->ss_state.search.delv_bins;
+    int64_t* bin_end = &cc->ss_state.search.bin_end;
+    uint32_t* bin_time = &cc->ss_state.search.bin_time;
+    uint16_t* bin_rounds = &cc->ss_state.search.bin_rounds;
 
-	// bin time is the size of each of the sent/delv bins
-  uint32_t tmp_bin_time = (loss->rtt.latest * QUICLY_SEARCH_WINDOW_MULTIPLIER) / (QUICLY_SEARCH_DELV_BIN_COUNT);
-  *bin_time = tmp_bin_time < 1 ? 1 : tmp_bin_time;
-	*bin_end = now + *bin_time;
-	delv[0] = 0;
-	*bin_rounds = 0;
+    // bin time is the size of each of the sent/delv bins
+    uint32_t tmp_bin_time = (loss->rtt.latest * QUICLY_SEARCH_WINDOW_MULTIPLIER) / (QUICLY_SEARCH_DELV_BIN_COUNT);
+    *bin_time = tmp_bin_time < 1 ? 1 : tmp_bin_time;
+    *bin_end = now + *bin_time;
+    delv[0] = bytes;
+    *bin_rounds = 0;
 }
 
 // bytes is the number of bytes acked in the last ACK frame
 // inflight is sentmap->bytes_in_flight + bytes
 void ss_search(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, uint64_t largest_acked, uint32_t inflight,
                         uint64_t next_pn, int64_t now, uint32_t max_udp_payload_size)
 {
-	// Handy pointers to the cc struct
-	uint64_t* delv = cc->ss_state.search.delv_bins;
-	int64_t* bin_end = &cc->ss_state.search.bin_end;
-	uint32_t* bin_time = &cc->ss_state.search.bin_time;
-	uint32_t* bin_rounds = &cc->ss_state.search.bin_rounds;
+    // Handy pointers to the cc struct
+    QUICLY_SEARCH_BIN_TYPE* delv = cc->ss_state.search.delv_bins;
+    int64_t* bin_end = &cc->ss_state.search.bin_end;
+    uint32_t* bin_time = &cc->ss_state.search.bin_time;
+    uint16_t* bin_rounds = &cc->ss_state.search.bin_rounds;
+
+    // struct initializations, everything else important has already been reset to 0
+    if(*bin_time == 0) {
+        ss_search_reset(cc, loss, bytes, now);
+    }
 
-	// struct initializations, everything else important has already been reset to 0
-	if(*bin_time == 0) {
-		ss_search_reset(cc, loss, bytes, now);
-	}
+    // perform a reset if a certain number of bins were missed (application limited case).
+    // non-application-limited delay will trigger a protocol reset if in-flight packets
+    // are not acked (typical case, not handled by us).
+    if (((now - *bin_end) / *bin_time) >= QUICLY_SEARCH_RESET_LIMIT) {
+        ss_search_reset(cc, loss, bytes + delv[(*bin_rounds % (QUICLY_SEARCH_TOTAL_BIN_COUNT))], now);
```

Comment:
This is intrinsic to SEARCH's ring buffer bin system, but this pattern of `bins[index % total bins]` comes up a lot.
In the FreeBSD implementation I defined a macro to reduce having to repeat that verbosely everywhere:
```c
#define SEARCH_BIN(ccv, index) ((struct newreno*)(ccv)->cc_data)->search_bin[(index) % SEARCH_TOTAL_BINS]
```

So the QUIC version could be something like
```c
#define SEARCH_BIN(delv, index) delv[(index) % (QUICLY_SEARCH_TOTAL_BIN_COUNT)]
```
Just thought I'd suggest it since this pattern shows up a lot and could easily become the location of a memory safety bug if not careful in review!

Reply:
Absolutely, that is something I had been thinking about as well. I'll give it a shot, it'll be nice to have that cleaned up a little.